pkg/middleware/stored_session.go

```diff
@@ -108,49 +108,47 @@ func (s *storedSessionLoader) getValidatedSession(rw http.ResponseWriter, req *h
 
 // refreshSessionIfNeeded will attempt to refresh a session if the session
 // is older than the refresh period.
-// It is assumed that if the provider refreshes the session, the session is now
-// valid.
-// If the session requires refreshing but the provider does not refresh it,
-// we must validate the session to ensure that the returned session is still
-// valid.
+// Success or fail, we will then validate the session.
 func (s *storedSessionLoader) refreshSessionIfNeeded(rw http.ResponseWriter, req *http.Request, session *sessionsapi.SessionState) error {
 	if s.refreshPeriod <= time.Duration(0) || session.Age() < s.refreshPeriod {
 		// Refresh is disabled or the session is not old enough, do nothing
 		return nil
 	}
 
 	logger.Printf("Refreshing %s old session cookie for %s (refresh after %s)", session.Age(), session, s.refreshPeriod)
-	refreshed, err := s.refreshSessionWithProvider(rw, req, session)
+	err := s.refreshSession(rw, req, session)
 	if err != nil {
-		return err
+		// If a preemptive refresh fails, we still keep the session
+		// if validateSession succeeds.
+		logger.Errorf("Unable to refresh session: %v", err)
 	}
 
-	if !refreshed {
-		// Session wasn't refreshed, so make sure it's still valid
-		return s.validateSession(req.Context(), session)
-	}
-	return nil
+	// Validate all sessions after any Redeem/Refresh operation (fail or success)
+	return s.validateSession(req.Context(), session)
```

Comment:
Session validation in some instances calls the API/token endpoint, this would cause an increase in the number of times we are hitting the API, anything we can do about this?

Reply:
Yeah, that was my concern about this too 🤔 

Not sure on that one, some providers send a dummy request to an endpoint to verify the access token (which probably isn't needed if a refresh just granted the new token -- but maybe doesn't hurt to make sure it work?).

The other case we will have more of in OIDC providers is verifying security-based IDToken claims - so their implementation won't have the API request.

In practice at the moment, the only provider's that implemented an API reqeust Validation didn't implement a RefreshSession function, so nothing currently has a double hit 🤷‍♂️

Reply:
This is still an issue.

Is it worth checking on Google/Azure to see if `ValidateSession` can be refactored to check the IDToken validity like OIDC instead of a Profile URL network request?

Or is 2 IdP round trips instead of 1 on refreshes not too big of a deal?

CC: @weinong @codablock

Reply:
I'm not sure what to do on this one, if it is just the two providers, perhaps we can look into deduplicating it, but  you're right, 2 calls may not be the end of the world

Reply:
Ideally, we get Azure & Google as OIDC based providers long term. Then we validate with an IDToken signature verification instead of a profileURL API call.

Reply:
What do you want to do then with regards to this PR for now?